
<style type="text/css">

<!--==========================================================================
	FTL macros 
	========================================================================== -->


<!--==========================================================================
	CSS styles 
	========================================================================== -->


<style type="text/css">

body {
	background-color: #dddddd;
}

/* Table styles. */

table {
  border-color: #000;
  border-spacing: 0px;
  border-style: solid;
  border-width: 1px;
  cell-spacing: 0px;
}

.noBorder {
  border-width: 0px;
}

td, th {
  font-family: Arial, Helvetica, sans-serif;
  font-size: 10pt;
  padding: 2px 0.5em;
  white-space: nowrap;
}

td.numeric {
  text-align: right;
}

th {
  background-color: #c0c0c0;
}

th.mainHeader {
  background-color: #808080;
  color: #ffffff;
  text-align: left;
}

th a {
  color: #000080;
  text-decoration: none;
}

th a:visited {
  color: #000080;
}

th a:active, th a:hover {
  color: #800000;
  text-decoration: underline;
}

.toc {
	border: 1px solid #aaa;
	background-color: #eeeeee;
	padding: 5px;
	font-size: 95%;
}

.note {
	border: 1px solid #aaa;
	background-color: #eeeeee;
	padding: 5px;
	font-size: 90%;
	text-align: left;
}

.main {
	margin-top: 15px;
	width: 1000px;
	text-align: left;
	background: #ffffff;
	border: 3px solid #777777;
}

.histo {
	table-layout:fixed;
	width:100%;
	border:1px solid #aaa;
	word-wrap:break-word;
}

</style>


<!--==========================================================================
	Index
	========================================================================== -->

<center>
<div class="main">

<center> <h3> SnpEff: Variant analysis </h3> </center>

<div style="margin-left: .5em">
<table class="toc"><tr><td>
	<center><b>Contents</b></center>
	<a href="#summary">Summary</a><br>
	<a href="#changeRateChr"> Change rate by chromosome</a><br>
	<a href="#changesByType">Variants by type</a><br>
	<a href="#effectsImpact"> Number of variants by impact </a><br>
	<a href="#effectsImpact"> Number of variants by functional class </a><br>
	<a href="#effects"> Number of variants by effect </a><br>
	<a href="#quality">Quality histogram</a><br>
	<a href="#coverage">Coverage histogram</a><br>
	<a href="#baseChages">Base change table</a><br>
	<a href="#tstv">Transition vs transversions (ts/tv)</a><br>
	<a href="#alleleFreq"> Frequency of alleles </a><br>
	<a href="#codonChanges"> Codon change table </a><br>
	<a href="#aaChanges"> Amino acid change table </a><br>
	<a href="#chrChanges"> Chromosome change plots </a><br>
	<a href="snpEff_genes.txt"> Details by gene </a><br>
</tr></td></table>
</div>


<!--==========================================================================
	Summary table 
	========================================================================== -->

<hr>
<a name="summary">
<center>
<b>Summary</b><p>

<table border=0>
	<tr bgcolor=ffffff> 
		<td valign=top> <b> Genome </b> </td>
		<td> egrandis201 </td>
	</tr>
	<tr bgcolor=dddddd> 
		<td valign=top> <b> Date </b> </td>
		<td> 2012-12-02 14:52 </td>
	</tr>
	<tr bgcolor=ffffff> 
		<td valign=top> <b> SnpEff version </b> </td>
		<td> <pre>SnpEff 3.1f (build 2012-12-01), by Pablo Cingolani</pre> </td>
	</tr>
	<tr bgcolor=dddddd> 
		<td valign=top> <b> Command line arguments </b> </td>
		<td> <pre>SnpEff  -v egrandis201 /home/pcingola/snpEff/pooja.vcf 
</pre> </td>
	</tr>
	<tr bgcolor=ffffff> 
		<td valign=top> <b> Warnings </b> </td>
		 
		<td bgcolor="#ffffff"> 0 </td>
	</tr>	
	<tr bgcolor=dddddd>
		<td valign=top> <b> Number of lines (input file) </b> </td>
		<td> 1 </td>
	</tr>
	<tr bgcolor=ffffff>
		<td valign=top> <b> Number of variants (before filter) </b> </td>
		<td> 1 </td>
	</tr>
	<tr bgcolor=dddddd>
		<td valign=top> <b> Filter </b> </td>
		<td>  </td>
	</tr>
	<tr bgcolor=dddddd>
		<td valign=top> <b> Number of variants filtered out </b> </td>
		<td> 0 </td>
	</tr>
	<tr bgcolor=ffffff>
		<td valign=top> <b> Number of not variants <br>(i.e. reference equals alternative) </b> </td>
		<td> 0 </td>
	</tr>
	<tr bgcolor=dddddd>
		<td valign=top> <b> Number of variants processed <br> (i.e. after filter and non-variants) </b> </td>
		<td> 1 </td>
	</tr>
	<tr bgcolor=ffffff>
		<td valign=top> <b> Number of known variants <br>(i.e. non-empty ID) </b> </td>
		<td> 
		0
		( 0% ) 
		</td>
	</tr>
	<tr bgcolor=ffffff>
		<td valign=top> <b> Number of effects </b> </td>
		<td> 2 </td>
	</tr>
	<tr bgcolor=dddddd>
		<td valign=top> <b> Genome total length </b> </td>
		<td> 30,869 </td>
	</tr>
	<tr bgcolor=ffffff>
		<td valign=top> <b> Genome effective length </b> </td>
		<td> 28,248 </td>
	</tr>
	<tr bgcolor=dddddd>
		<td valign=top> <b> Change rate </b> </td>
		<td> 1 change every 28,248 bases </td>
	</tr>
</table>
<p>
</center>

<!--==========================================================================
	Change rate by chromosome
	========================================================================== -->

<hr> 
<a name="changeRateChr"> 

<center>
<b> Change rate details </b><p>

<table border=1>
	<tr><th> Chromosome </th><th> Length </th><th> Changes </th><th> Change rate </th></tr> 
		<tr>
			<td> 441 </td>
			<td class="numeric"> 28,248 </td>
			<td class="numeric"> 1 </td>
			<td class="numeric"> 28,248 </td>
		</tr> 
	<tr>
		<th> Total </th>
		<th class="numeric"> 28,248 </th>
		<th class="numeric"> 1 </th>
		<th class="numeric"> 28,248 </th>
	</tr> 
</table> 
</center>

<!--==========================================================================
	Changes by type
	========================================================================== -->

<hr> 
<a name="changesByType"> 
<center>
<b> Number changes by type</b><p>

<table border=1>
	<thead>
		<tr>
			<th> <b> Type   </b> </th>
			<th> <b> Total  </b> </th>
			<th> <b> Homo   </b> </th>
			<th> <b> Hetero </b> </th>
		</tr>
	</thead>
    <tbody>
	    <tr>
	    	<td> <b> SNP </b> </td>
	    	<td class="numeric" bgcolor="ffffff"> 1 </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    	<td class="numeric" bgcolor="ffffff"> 1 </td>
	    </tr>
	    <tr>
	    	<td> <b> MNP </b> </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    </tr>
	    <tr>
	    	<td> <b> INS </b> </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    </tr>
	    <tr>
	    	<td> <b> DEL </b> </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    </tr>
	    <tr>
	    	<td> <b> MIXED </b> </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    </tr>
	    <tr>
	    	<td> <b> Interval </b> </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    	<td class="numeric" bgcolor="ffffff"> 0 </td>
	    </tr>
    </tbody>
    <tfoot>
	    <tr>
	    	<th><b>Total </b> </th>
	    	<th class="numeric"> 1 </th>
	    	<th class="numeric"> 0 </th>
	    	<th class="numeric"> 1 </th>
	    </tr>
    </tfoot>
</table>
</center>

<!--==========================================================================
	Effects by impact
	========================================================================== -->

<hr> 
<a name="effectsImpact"> 
<center>
<b> Number of effects by impact </b> <p> 

<table border=0>
	<thead>
		<tr> 
			<th><b> Type (alphabetical order)  </b></th>
			<th> &nbsp; </th>
			<th> Count </th>
			<th> Percent </th>
		</tr>
	</thead>
		<tr> 
			<td> <b> LOW </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="ffffff"> 1 </td> 
			<td class="numeric" bgcolor="ffffff"> 50% </td>
		</tr>
		<tr> 
			<td> <b> MODIFIER </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="ffffff"> 1 </td> 
			<td class="numeric" bgcolor="ffffff"> 50% </td>
		</tr>
</table><br>
<p>
</center>

<!--==========================================================================
	Effects by functional class
	========================================================================== -->

<hr> 
<a name="effectsFuncClass"> 
<center>
<b> Number of effects by functional class </b> <p> 

<table border=0>
	<thead>
		<tr> 
			<th><b> Type (alphabetical order)  </b></th>
			<th> &nbsp; </th>
			<th> Count </th>
			<th> Percent </th>
		</tr>
	</thead>
		<tr> 
			<td> <b> SILENT </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="ffffff"> 1 </td> 
			<td class="numeric" bgcolor="ffffff"> 100% </td>
		</tr>
</table><br>
<p>

Missense / Silent ratio: </th><td class="numeric"> 0

</center>

<!--==========================================================================
	Effects stats
	========================================================================== -->

<hr> 
<a name="effects"> 
<center>
<b> Number of effects by type and region </b> <p> 

<table border=0>
	<tr>
		<th> Type </th>
		<th> Region </th>
	</tr>
	<tr>
		<td> <table border=0>
	<thead>
		<tr> 
			<th><b> Type (alphabetical order)  </b></th>
			<th> &nbsp; </th>
			<th> Count </th>
			<th> Percent </th>
		</tr>
	</thead>
		<tr> 
			<td> <b> DOWNSTREAM </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="ffffff"> 1 </td> 
			<td class="numeric" bgcolor="ffffff"> 50% </td>
		</tr>
		<tr> 
			<td> <b> SYNONYMOUS_CODING </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="ffffff"> 1 </td> 
			<td class="numeric" bgcolor="ffffff"> 50% </td>
		</tr>
</table><br>
 </td>
		<td> <table border=0>
	<thead>
		<tr> 
			<th><b> Type (alphabetical order)  </b></th>
			<th> &nbsp; </th>
			<th> Count </th>
			<th> Percent </th>
		</tr>
	</thead>
		<tr> 
			<td> <b> DOWNSTREAM </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="ffffff"> 1 </td> 
			<td class="numeric" bgcolor="ffffff"> 50% </td>
		</tr>
		<tr> 
			<td> <b> EXON </b> </td> 
			<th> &nbsp; </th>
			<td class="numeric" bgcolor="ffffff"> 1 </td> 
			<td class="numeric" bgcolor="ffffff"> 50% </td>
		</tr>
</table><br>
 </td>
	</tr>
</table>
<p>

<img src="http://chart.apis.google.com/chart?cht=bvs&chbh=50,4,8&chs=800x300&chxt=y,y,x,x&chtt=Variations&chxp=1,50.0|3,50.0&chxr=0,0.0,50.0|1,0.0,100.0|3,0.0,100.0&chco=00FFFF,FA8072,FF0000,A52A2A,FFA500,DEB887,E9967A,7FFFD4,808080&chdl=Upstream|5%27UTR|Exon|Splice+Donor|Intron|Splice+Acceptor|3%27UTR|Downstream|Intergenic&chxl=1:|%25|2:|Intergenic|Up|5%27UTR|Exon|Donor|Intron|Acceptor|Exon|3%27UTR|Down|Intergenic|3:|&chd=e:AAAAAAAAAAAAAAAAAAAAAA,AAAAAAAAAAAAAAAAAAAAAA,AAAAAA..AAAAAA..AAAAAA,AAAAAAAAAAAAAAAAAAAAAA,AAAAAAAAAAAAAAAAAAAAAA,AAAAAAAAAAAAAAAAAAAAAA,AAAAAAAAAAAAAAAAAAAAAA,AAAAAAAAAAAAAAAAAA..AA,AAAAAAAAAAAAAAAAAAAAAA" border=1><p>

</center>
<!--==========================================================================
	Quality plots
	========================================================================== -->

	<hr> 
	<a name="quality"> 
	<b> Quality: </b> 
	<p>
	<pre>
	<table class="histo">
		<tr> <th width=15%>Min</th><td>111</td> </tr>
		<tr> <th>Max</th><td>111</td> </tr>
		<tr> <th>Mean</th><td>111</td> </tr>
		<tr> <th>Median</th><td>55.5</td> </tr>
		<tr> <th>Standard deviation</th><td>0</td> </tr>
		<tr> <th>Values</th><td>111	</td> </tr>
		<tr> <th>Count</th><td>1	</td> </tr>
		</tr>
	</table>
		<img src="http://chart.apis.google.com/chart?cht=bvg&chd=e:..&chbh=8,4,2&chs=800x300&chxt=y,y,x,x&chxp=1,50.0|3,50.0&chtt=Quality+histogram&chxl=1:|Count|2:|111|3:|Quality&chxr=0,0.0,1.0|1,0.0,100.0|3,0.0,100.0"><br>
	</pre>

<!--==========================================================================
	Coverage
	========================================================================== -->

	<hr> 
	<a name="coverage"> 
	<b> Coverage:</b> 
	<p>
	<pre>
	<table class="histo">
		<tr> <th width=15%>Min</th><td>35</td> </tr>
		<tr> <th>Max</th><td>35</td> </tr>
		<tr> <th>Mean</th><td>35</td> </tr>
		<tr> <th>Median</th><td>17.5</td> </tr>
		<tr> <th>Standard deviation</th><td>0</td> </tr>
		<tr> <th>Values</th><td>35	</td> </tr>
		<tr> <th>Count</th><td>1	</td> </tr>
		</tr>
	</table>
		<img src="http://chart.apis.google.com/chart?cht=bvg&chd=e:..&chbh=8,4,2&chs=800x300&chxt=y,y,x,x&chxp=1,50.0|3,50.0&chtt=Coverage+histogram&chxl=1:|Count|2:|35|3:|Coverage&chxr=0,0.0,1.0|1,0.0,100.0|3,0.0,100.0"><br>
	</pre>

<!--==========================================================================
	InDels
	========================================================================== -->


<!--==========================================================================
	Base changes
	========================================================================== -->

<hr> 
<a name="baseChages"> 
<center>
<b> Base changes (SNPs) </b> <p>

<table border=1>
    <tr>
    	<td> &nbsp; </td>  <th> <b> A </b> </th> <th> <b> C </b> </th> <th> <b> G </b> </th> <th> <b> T </b> </th>
    </tr>
	    <tr> <th> <b> A </b> </th><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td> </tr>
	    <tr> <th> <b> C </b> </th><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td> </tr>
	    <tr> <th> <b> G </b> </th><td class="numeric" bgcolor="ffffff"> 1 </td><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td> </tr>
	    <tr> <th> <b> T </b> </th><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td><td class="numeric" bgcolor="ffffff"> 0 </td> </tr>
</table>
</center>

<p>

<!--==========================================================================
	Ts/Tv
	========================================================================== -->

<hr> 
<a name="tstv">
<center> <b> Ts/Tv (transitions / transversions) </b> <p> </center>

<small>
<b>Note:</b> Only SNPs are used for this statistic.<br>
<b>Note:</b> This Ts/Tv ratio is a 'raw' ratio. Some people prefer to use a ratio of rates, not observed events. In that case, you need to multiply by 2.0 (since there are twice as many possible transitions than transversions, E[Ts/Tv] ratio is twice the ratio of events).
</small>
<p>

<center>
<table border=1>
	<tr> <th> Transitions </th><td class="numeric"> 1 </td> </tr>
	<tr> <th> Transversions </th><td class="numeric"> 0 </td> </tr>
	<tr> <th> Ts/Tv ratio </th><td class="numeric"> 0 </td> </tr>
</table>
</center>
<p>

	<b>All variants:</b>
	<pre>Sample        :	Total
Transitions   :	1	1
Transversions :	0	0
Ts/Tv         :	Infinity	Infinity	
</pre>
	<p>
	<b>Only known variants</b> (i.e. the ones having a non-empty ID field):
	<pre>No results available (empty input?)</pre>
<p>

<!--==========================================================================
	Allele frequency
	========================================================================== -->

<hr> 
<a name="alleleFreq">
<center> <b> Frequency of alleles </b> <p> </center>

Note: Number of times an allele appears once (singleton), twice (doubletons), etc.<p>

	<b>All variants:</b>
	<img src="http://chart.apis.google.com/chart?cht=bvg&chd=e:..&chbh=8,4,2&chs=800x300&chxt=y,y,x,x&chxp=1,50.0|3,50.0&chtt=Alleles+frequencies&chxl=1:|Count|2:|1|3:|Alleles&chxr=0,0.0,1.0|1,0.0,100.0|3,0.0,100.0"><br>
	<table class="histo">
		<tr> <th width=15%>Min</th><td>1</td> </tr>
		<tr> <th>Max</th><td>1</td> </tr>
		<tr> <th>Mean</th><td>1</td> </tr>
		<tr> <th>Median</th><td>0.5</td> </tr>
		<tr> <th>Standard deviation</th><td>0</td> </tr>
		<tr> <th>Values</th><td>1	</td> </tr>
		<tr> <th>Count</th><td>1	</td> </tr>
		</tr>
	</table>
	<p>
	<b>Only known variants</b> (i.e. the ones having a non-empty ID field):
	<table class="histo">
		<tr> <th width=15%>Min</th><td>2,147,483,647</td> </tr>
		<tr> <th>Max</th><td>-2,147,483,648</td> </tr>
		<tr> <th>Mean</th><td>0</td> </tr>
		<tr> <th>Median</th><td>0</td> </tr>
		<tr> <th>Standard deviation</th><td>0</td> </tr>
		<tr> <th>Values</th><td></td> </tr>
		<tr> <th>Count</th><td></td> </tr>
		</tr>
	</table>
<p>


<!--==========================================================================
	Codon change table
	========================================================================== -->
	
<hr> 
<a name="codonChanges">
<center>
<b> Codon changes</b> <p>

	<div class="note">
		How to read this table: <br>
		- Rows are reference codons and columns are changed codons. E.g. Row 'AAA' column 'TAA' indicates how many 'AAA' codons have been replaced by 'TAA' codons.<br>
		- Red background colors indicate that more changes happened (heat-map).<br>
		- Diagonals are indicated using grey background color <br> 
		- WARNING: This table may include different translation codon tables (e.g. mamalian DNA and mitochondrial DNA).<br>
		<p>
	</div><p>

  
<table border=1>
	<thead>
		<tr> 
			<th> &nbsp; </th>
			 <th> TAC </th>  <th> TAT </th> 
		</tr>
	</thead>
			<tr> 
				<th> TAC </th>
					
						<th class="numeric"> &nbsp </th>
					
						<td class="numeric" bgcolor="ffffff"> 1 </td>
			</tr>
			<tr> 
				<th> TAT </th>
					
						<td class="numeric" bgcolor="ffffff"> &nbsp </td>
					
						<th class="numeric"> &nbsp </th>
			</tr>
</table>
</center>

<!--==========================================================================
	Amino acid change table
	========================================================================== -->

<hr> 
<a name="aaChanges">
<center>
<b> Amino acid changes</b><p> 

	<div class="note">
		How to read this table: <br>
		- Rows are reference amino acids and columns are changed amino acids. E.g. Row 'A' column 'E' indicates how many 'A' amino acids have been replaced by 'E' amino acids.<br>
		- Red background colors indicate that more changes happened (heat-map).<br>
		- Diagonals are indicated using grey background color <br> 
		- WARNING: This table may include different translation codon tables (e.g. mamalian DNA and mitochondrial DNA).<br>
		<p>
	</div><p>

<table border=1>
	<thead>
		<tr> 
			<th> &nbsp; </th>
			 <th> Y </th> 
		</tr>
	</thead>
			<tr> 
				<th> Y </th>
					
						<th class="numeric"> 1 </th>
			</tr>
</table>
</center>

<!--==========================================================================
	Chromosome change table
	========================================================================== -->

<hr> 
<a name="chrChanges">
<center>
<b> Changes by chromosome</b><p> 

<center>
	<pre>
		<img src="http://chart.apis.google.com/chart?cht=lc&chs=800x300&chxt=y,y,x,x&chxp=1,50.0|3,50.0&chtt=Changes+histogram%3A+441&chxr=0,0.0,0.0|1,0.0,100.0|3,0.0,100.0&chxl=1:|Changes%2F100b|2:|150||||||||||3700||||||||||7250||||||||||10800||||||||||14350||||||||||17850||||||||||21400||||||||||24950|||||||||28100|3:|Position&chd=e:AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"><br>
		Chromosome: 441
	Position :	0	100	200	300	400	500	600	700	800	900	1000	1100	1200	1300	1400	1500	1600	1700	1800	1900	2000	2100	2200	2300	2400	2500	2600	2700	2800	2900	3000	3100	3200	3300	3400	3500	3600	3700	3800	3900	4000	4100	4200	4300	4400	4500	4600	4700	4800	4900	5000	5100	5200	5300	5400	5500	5600	5700	5800	5900	6000	6100	6200	6300	6400	6500	6600	6700	6800	6900	7000	7100	7200	7300	7400	7500	7600	7700	7800	7900	8000	8100	8200	8300	8400	8500	8600	8700	8800	8900	9000	9100	9200	9300	9400	9500	9600	9700	9800	9900	10000	10100	10200	10300	10400	10500	10600	10700	10800	10900	11000	11100	11200	11300	11400	11500	11600	11700	11800	11900	12000	12100	12200	12300	12400	12500	12600	12700	12800	12900	13000	13100	13200	13300	13400	13500	13600	13700	13800	13900	14000	14100	14200	14300	14400	14500	14600	14700	14800	14900	15000	15100	15200	15300	15400	15500	15600	15700	15800	15900	16000	16100	16200	16300	16400	16500	16600	16700	16800	16900	17000	17100	17200	17300	17400	17500	17600	17700	17800	17900	18000	18100	18200	18300	18400	18500	18600	18700	18800	18900	19000	19100	19200	19300	19400	19500	19600	19700	19800	19900	20000	20100	20200	20300	20400	20500	20600	20700	20800	20900	21000	21100	21200	21300	21400	21500	21600	21700	21800	21900	22000	22100	22200	22300	22400	22500	22600	22700	22800	22900	23000	23100	23200	23300	23400	23500	23600	23700	23800	23900	24000	24100	24200	24300	24400	24500	24600	24700	24800	24900	25000	25100	25200	25300	25400	25500	25600	25700	25800	25900	26000	26100	26200	26300	26400	26500	26600	26700	26800	26900	27000	27100	27200	27300	27400	27500	27600	27700	27800	27900	28000	28100	28200	
	Count    :	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	1	0	0	0	0	

	</pre>
</center>

<!--==========================================================================
	Changes by gene 
	========================================================================== -->

<center>
	<hr>
	<b> Details by gene </b><p>
	<p>
	<b><a href="snpEff_genes.txt">Here</a></b> you can find a tab-separated table.
</center>
</div>
</center>