
	<div id="content">
		&nbsp;<p>

		This project is maintained by <b> <a class="body" href="http://www.mcb.mcgill.ca/~pcingola/">Pablo Cingolani</a> (pcingola@users.sourceforge.net</a>)</b><p>

		<b> Acknowledgements:</b> <br> 
		<ul>
			<li> Adrian Platts (McGill): tons of feedback, feature suggestions, test cases, etc.<br>
			<li> Doug Ruden (WSU): feedback, features request, debugging.<br>
			<li> Dave OConnor (UW-Madison): feature suggestions, HIV genome <br>
			<li> Denis Reshetov (Rogaev lab): VCF output support and debugging on GTF parsing.<br>
			<li> Eric Banks & David Rozen (Broad): VCF parsing and VCF output.<br>
			<li> Jonathan Fresnedo (University of California, Davis): P.Persica genome support <br>
			<li> <b>GATK:</b> integratetion was performed by Broad's Genome Analysis Team: David Rozen, Eric Banks and Mark DePristo
			<li> Arno Velds (NKI: Netherlands Cancer Institute): Annomalities in ENSEMBL annotation files that led to '-onlyCoding = auto' development, as well as "rogue transcript filters". <br>
			<li> Giulio Genovese: ENSEMBL annotation problems, fixing and improving "rogue transcript" problem. <br>
			<li> Louis Letourneau: Maven project
			<li> Peter Briggs (Manchester University) : Improved Galaxy interface and wrappers. <br>
			<li> Jim Johnson (Minnesota Supercomputing Institute, University of Minnesota): SnpSift improvements and bugfixes. <br>
			<li> Davide Cittaro (Center for Translational Genomics and Bioinformatics, Ospedale San Raffaele): Debugging and beta-testing for both SnpEff & SnpSift. Helped to debug many interoperability, compatibility and format specification issues. <br>
		</ul>
		<p>

		<p style="font-size: 8pt" align=center><font color=#777777> - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - </font></p>

		Please send any bug reports to <b>pcingola at users.sourceforge.net</b>.<br>
		<b>Note:</b> A bug report should include:
		<pre>
			- An explanation of the bug condition

			- A minimal bug-demonstrating test case (i.e. include SNP file used to run the test case)

			- The conditions for the bug to occur (e.g. genome version)

			- snpEff version you are using
		</pre>
		&nbsp;<p>
		&nbsp;<p>
		&nbsp;<p>
		&nbsp;<p>
		&nbsp;<p>
		&nbsp;<p>
		&nbsp;<p>
		&nbsp;<p>
		&nbsp;<p>
		&nbsp;<p>
		&nbsp;<p>
		&nbsp;<p>
		&nbsp;<p>
		&nbsp;<p>
	</div>
